dali/operators/decoder/audio/utils.h

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_DECODER_AUDIO_UTILS_H_
+#define DALI_OPERATORS_DECODER_AUDIO_UTILS_H_
+
+#include <functional>
+#include <numeric>
+#include <utility>
+#include <vector>
+#include <cassert>
+#include <cmath>
+#include "dali/core/error_handling.h"
+#include "dali/core/span.h"
+
+namespace dali {
+
+/**
+ * Downmixing data to mono-channel. It's safe to do it in-place.
+ *
+ * Assumes, that `out` buffer is properly allocated
+ * (i.e. has `n_in`/`weights.size()` elements at least)
+ *
+ * @param weights Specifies weight for every channel.
+ */
+template<typename T>
+void Downmixing(T *out, const T *in, size_t n_in, const std::vector<int> &weights) {
+  DALI_ENFORCE(out && in, "Incorrect input or output buffer. Or both.");
+  DALI_ENFORCE(weights.size() > 1, "You can't downmix mono-channel data");
+  DALI_ENFORCE(n_in % weights.size() == 0, "Data layout incorrect");
+  for (size_t i = 0; i < n_in; i++) {
+    *out = (*in++) * weights[0];
+    for (size_t j = 1; j < weights.size(); j++) {
+      *out += (*in++) * weights[j];
+    }
+    *out++ /= std::accumulate(weights.begin(), weights.end(), 0);
```

Comment:
1. Pointer arithmetic should rather be avoided (indexing is much more readable)
2. The normalization factor can be hoisted outside the loop.
3. Do we even _want_ this normalization? When operating on floating point types, we might exceed -1..1 range to no ill effects. On integral types, we may prefer to have occasional clipping and/or compression instead of making most of the audio quieter and therefore exacerbating quantization noise.
4. The best way to normalize (if you insist on that, but I'd make it an option) would be to have a local copy of the weights and normalize _them_, avoiding extra arithmetic for each sample.
5. This would be a very good place to change the type from native (e.g. int16_t) to float.

I was thinking about sth like this:
```
template <typename Out, typename In>
void Downmix(Out *out, const In *in, int64_t samples, int channels, const float *weights, bool normalize_weights = false) {
  SmallVector<float, 8> normalized_weights;  // 8 channels should be enough for 7.1 audio
  if (normalize_weights) {
    double sum = 0;
    for (int i = 0; i < channels; i++)
      sum += weights[i];
    normalized_weights.resize(channels);
    for (int i = 0; i < channels; i++) {
      normalized_weights[i] = weights[i] / sum;
    }
    weights = normalized_weights.data();  // use this pointer now
  }
  for (int64_t o = 0, i = 0; o < samples; o++, i += channels) {
    float sum = ConvertNorm<float>(in[i]) * weights[0];  // save one addition per sample
    for (int c = 1; c < channels; c++) {
      sum += ConvertNorm<float>(in[i + c]) * weights[1];
    }
    out[o] = ConvertSatNorm<Out>(sum);
  }
}
```